File: alternative/medium/clone_graph.py
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

from typing import Optional

def cloneGraph(node: Optional['Node']) -> Optional['Node']:
    if not node:
        return None

    visited = {}

    def dfs(node):
        if node in visited:
            return visited[node]
        clone = Node(node.val, [])
        visited[node] = clone
        for n in node.neighbors:
            clone.neighbors.append(dfs(n))
        return clone

    return dfs(node)

File: src/medium/clone_graph.rs
#![allow(dead_code)]
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

#[derive(Clone)]
pub struct Node {
    pub val: i32,
    pub neighbors: Vec<Rc<RefCell<Node>>>,
}

impl Node {
    #[inline]
    pub fn new(val: i32, neighbors: Vec<Rc<RefCell<Node>>>) -> Self {
        Node { val, neighbors }
    }
}

pub fn clone_graph(node: Option<Rc<RefCell<Node>>>) -> Option<Rc<RefCell<Node>>> {
    fn dfs(
        node: Rc<RefCell<Node>>,
        visited: &mut HashMap<i32, Rc<RefCell<Node>>>,
    ) -> Rc<RefCell<Node>> {
        if let Some(v) = visited.get(&node.borrow().val) {
            return v.clone();
        }
        let clone = Rc::new(RefCell::new(Node::new(node.borrow().val, vec![])));
        visited.insert(node.borrow().val, clone.clone());
        for n in &node.borrow().neighbors {
            clone.borrow_mut().neighbors.push(dfs(n.clone(), visited));
        }
        clone
    }

    let mut visited = HashMap::new();
    node.map(|n| dfs(n, &mut visited))
}
